Automatic R&D sync

scriptis opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This PR adds hooks to techfabs and exofabs to automatically synchronize research when the linked techweb updates, and deletes the old R&D sync buttons. It also lowers `TGUI_REFRESH_FULL_UPDATE_COOLDOWN` to one second (one fifth of its original value), and now actually uses that value to compute when to send a full refresh.

Prior to this, windows would hang on "please wait" for 5-10 seconds, which felt _really_ broken (is this even doing anything?).

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

People will stop asking me to add this functionality, and people will no longer close out of tgui windows because they think they're broken when really they have to wait eight seconds for a refresh

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
qol: techfabs, protolathes, circuit imprinters, and exosuit fabricators now automatically r&d sync
qol: the full window refresh time is now one second instead of five seconds
fix: refreshing windows ("please wait...") no longer hang for up to twice as long as they should
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Commit:
Automatic R&D sync

Commit:
One second instead of a half-second

JohnFulpWillard commented on code/modules/tgui/tgui.dm:
```diff
@@ -203,9 +203,8 @@
 	if(!COOLDOWN_FINISHED(src, refresh_cooldown))
 		refreshing = TRUE
-		addtimer(CALLBACK(src, .proc/send_full_update), TGUI_REFRESH_FULL_UPDATE_COOLDOWN, TIMER_UNIQUE)
+		addtimer(CALLBACK(src, .proc/send_full_update), COOLDOWN_TIMELEFT(src, refresh_cooldown), TIMER_UNIQUE)
```
If full update is called every second, why bother having a static data instead of just putting everything on regular ``ui_data``?

scriptis replied:
This timer isn't to invoke `send_full_update` every second--it's called specifically only when `refresh_cooldown` isn't done (e.g., cooldown not finished, check back when it is)

scriptis replied:
To clarify: static data shouldn't be getting sent every second. I don't think it is--unless I'm wrong?

JohnFulpWillard replied:
No you're fine, I just misread and thought it would constantly be calling this.

Mothblocks commented on code/modules/research/machinery/_production.dm:
```diff
@@ -43,6 +43,9 @@
 	update_icon(UPDATE_OVERLAYS)
 
+	RegisterSignal(stored_research, COMSIG_TECHWEB_ADD_DESIGN, .proc/update_designs)
+	RegisterSignal(stored_research, COMSIG_TECHWEB_REMOVE_DESIGN, .proc/update_designs)
```
Does update_designs have SIGNAL_HANDLER?

scriptis replied:
I'm blind, I'll get that in :)

Mothblocks commented on code/modules/vehicles/mecha/mech_fabricator.dm:
```diff
@@ -73,6 +73,8 @@
 	RefreshParts() //Recalculating local material sizes if the fab isn't linked
 	update_menu_tech()
+	RegisterSignal(stored_research, COMSIG_TECHWEB_ADD_DESIGN, .proc/update_menu_tech)
+	RegisterSignal(stored_research, COMSIG_TECHWEB_REMOVE_DESIGN, .proc/update_menu_tech)
```
Does this have signal handler?

Also for both of these just `RegisterSignal(x, list(signal1, signal2), .proc/update)`

Commit:
Moth changes + more

- Actually use SIGNAL_HANDLER
- Use update_static_data_for_all_viewers instead of only updating static data for the current user

Commit:
Batch techweb updates

scriptis commented:
CI being angry has nothing to do with me, this should be good to go unless you've got any other remarks :)
